### PR TITLE
CAPZ: Add AKS-based e2e workflow for ARO-HCP workload cluster lifecycle (ARO-27593)

### PR DESCRIPTION
**Pull request by marek-veber**
## Summary

Add a complete CAPZ e2e CI workflow under `Azure/ARO-HCP` that tests the full ARO-HCP workload cluster lifecycle via Cluster API on an AKS management cluster.

### How it works

1. **Provision ARO-HCP dev environment** — following the `aro-hcp-local-e2e` pattern, a fresh ARO-HCP environment is provisioned per job run (`aro-hcp-write-config` + `aro-hcp-provision-environment`). This deploys the full ARO-HCP stack (SVC cluster, management cluster, frontend, backend, etc.) and exposes a frontend endpoint URL.

2. **Connect to dev environment** (`capz-arohcp-dev-setup`) — runs in the `aro-hcp-e2e-tests` image (Go 1.25+, templatize). Gets the SVC cluster kubeconfig, discovers the frontend endpoint from the Istio VirtualService, grants frontend ingress for the CI runner, and registers the CI subscription. Writes `FRONTEND_ADDRESS` to `SHARED_DIR/dev_endpoint`.

3. **Provision AKS management cluster** (`capz-arohcp-aks-provision`) — creates a separate AKS cluster where CAPI/CAPZ/ASO operators will run.

4. **Deploy and test** — `cluster-api-installer` provides Helm charts and YAML generation (`gen.sh`) that configure CAPI/CAPZ/ASO with the correct managed identities. The generated CRs include an `HcpOpenShiftCluster` pointing at the dev frontend URL, enabling end-to-end testing of the full provisioning path without mocks.

5. **Cleanup** — tears down CAPZ resources, deprovisions the AKS cluster, collects ARO-HCP gather data, and deprovisions the ARO-HCP environment.

### Workflow steps

```
Pre:  aro-hcp-write-config → aro-hcp-provision-environment
      → aks-provision → check-dependencies → setup → dev-setup
Test: management-cluster → generate-yamls → cleanup-pre → deploy-crs
      → verify-workload-cluster → delete-workload-cluster → validate-cleanup
Post: summary → teardown → aks-deprovision
      → gather-provision-failure → gather-visualization
      → gather-custom-link-tools → gather-observability
      → deprovision-environment
```

### Jobs added

| Job name | Type | Schedule |
|----------|------|----------|
| `pull-ci-Azure-ARO-HCP-main-capz-e2e-dev` | presubmit | on PR (manual, `run_if_changed: ^$`) |
| `pull-ci-Azure-ARO-HCP-main-capz-e2e-images` | presubmit | on PR (auto, images only) |
| `periodic-ci-Azure-ARO-HCP-main-capz-e2e-production` | periodic | weekdays 20:00 UTC |

### Build configuration

- capi-tests image built inline via `dockerfile_literal` (kubectl, helm, clusterctl, gotestsum, azure-cli)
- `capz-test-env.sh` replaced with release-repo controlled version via `COPY <<'CAPZEOF'` heredoc
- 9 ARO-HCP images built for environment provisioning (e2e-tools, backend, frontend, admin-api, sessiongate, fleet, mgmt-agent, kube-applier, hcp-recovery)

### Managed identity (MSI) leasing

Leases 1 MSI container from Boskos pool `aro-hcp-test-msi-containers-<env>`. `LEASED_MSI_CONTAINERS` provides the resource group name used by `gen.sh` for identity-aware YAML generation with `detach-on-delete` reconcile policy.

## Test plan

- [x] CI rehearsal: full dev workflow passed (environment provisioning, AKS setup, dev-setup connection to frontend endpoint)
- [ ] CI rehearsal: full CAPZ test cycle (management-cluster through validate-cleanup)
- [ ] Periodic: production job on weekday schedule

Jira:
- [ARO-27593](https://issues.redhat.com/browse/ARO-27593) — CAPZ: Add AKS-based e2e workflow for ARO-HCP
- [ARO-25085](https://issues.redhat.com/browse/ARO-25085) — Test and document ARO-HCP devel environment integration with CAPZ e2e tests
- [ARO-27694](https://issues.redhat.com/browse/ARO-27694) — CAPZ: Testing plan for periodic ARO-HCP e2e tests on int, stg, prod environments

[ARO-27593]: https://redhat.atlassian.net/browse/ARO-27593?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse pull-ci-stolostron-capi-tests-configure-prow-mgmt-capz-e2e

File: ci-operator/config/stolostron/capi-tests/stolostron-capi-tests-main.yaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    requests:
+      cpu: 100m
+      memory: 200Mi
+tests:
+- as: capz-e2e
+  steps:
+    env:
+      ARO_REPO_BRANCH: ARO-27593-use-msi-ci-resources
+      ARO_REPO_DIR: /tmp/cluster-api-installer-pro
+      ARO_REPO_URL: https://github.com/marek-veber/cluster-api-installer.git
```
Comment (RadekCap) [inline ci-operator/config/stolostron/capi-tests/stolostron-capi-tests-main.yaml]: todo: marek -> stolostron

File: ci-operator/config/stolostron/capi-tests/stolostron-capi-tests-main__periodics.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      name: "4.19"
+      namespace: ocp
+resources:
+  '*':
+    requests:
+      cpu: 100m
+      memory: 200Mi
+tests:
+- as: capz-e2e
+  cron: 0 6 * * 1-5
```
Comment (RadekCap) [inline ci-operator/config/stolostron/capi-tests/stolostron-capi-tests-main__periodics.yaml]: Roughly about 2 hours before our current periodic if possible with coordination with Service Lifecycle team.

File: ci-operator/jobs/stolostron/capi-tests/stolostron-capi-tests-main-presubmits.yaml [hunk 3]
```diff
@@ -3,20 +3,18 @@ presubmits:
   - agent: kubernetes
     always_run: true
     branches:
-    - ^configure-prow-mgmt$
-    - ^configure-prow-mgmt-
-    cluster: build13
+    - ^main$
```
Comment (RadekCap) [inline ci-operator/jobs/stolostron/capi-tests/stolostron-capi-tests-main-presubmits.yaml]: This will run our test on each of our PRs into main.

**Comment by marek-veber** [comment]
/pj-rehearse branch-ci-Azure-ARO-HCP-main-capz-e2e-integration-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse branch-ci-Azure-ARO-HCP-main-capz-e2e-integration-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse periodic-ci-Azure-ARO-HCP-main-capz-e2e-production-capz-e2e

**Comment by marek-veber** [comment]
/pj-rehearse periodic-ci-Azure-ARO-HCP-main-capz-e2e-production

**Comment by marek-veber** [comment]
/pj-rehearse periodic-ci-Azure-ARO-HCP-main-capz-e2e-production

**Comment by marek-veber** [comment]
/pj-rehearse periodic-ci-Azure-ARO-HCP-main-capz-e2e-production

File: ci-operator/config/Azure/ARO-HCP/Azure-ARO-HCP-main__capz-e2e.yaml [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+tests:
+- as: dev
+  run_if_changed: ^$
+  steps:
+    env:
+      ARO_HCP_CLOUD: dev
+      ARO_HCP_DEPLOY_ENV: ci01
+      ARO_REPO_BRANCH: ARO-27593-use-msi-ci-resources
+      ARO_REPO_DIR: /tmp/cluster-api-installer-pro
+      ARO_REPO_URL: https://github.com/marek-veber/cluster-api-installer.git
```
Comment (RadekCap) [inline ci-operator/config/Azure/ARO-HCP/Azure-ARO-HCP-main__capz-e2e.yaml]: Use the upstream stolostron repo instead of the personal fork:

```suggestion
      ARO_REPO_BRANCH: aro-hcp-2025-23-12-preview
      ARO_REPO_DIR: /tmp/cluster-api-installer-pro
      ARO_REPO_URL: https://github.com/stolostron/cluster-api-installer.git
```

File: ci-operator/config/Azure/ARO-HCP/Azure-ARO-HCP-main__capz-e2e.yaml [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+    - ref: capz-arohcp-delete-workload-cluster
+    - ref: capz-arohcp-validate-cleanup
+  timeout: 4h0m0s
+- as: production
+  cron: 0 20 * * 1-5
+  steps:
+    env:
+      ARO_REPO_BRANCH: ARO-27593-use-msi-ci-resources
+      ARO_REPO_DIR: /tmp/cluster-api-installer-pro
+      ARO_REPO_URL: https://github.com/marek-veber/cluster-api-installer.git
```
Comment (RadekCap) [inline ci-operator/config/Azure/ARO-HCP/Azure-ARO-HCP-main__capz-e2e.yaml]: Same here — use the upstream stolostron repo:

```suggestion
      ARO_REPO_BRANCH: aro-hcp-2025-23-12-preview
      ARO_REPO_DIR: /tmp/cluster-api-installer-pro
      ARO_REPO_URL: https://github.com/stolostron/cluster-api-installer.git
```

File: ci-operator/config/Azure/ARO-HCP/Azure-ARO-HCP-main__capz-e2e.yaml [hunk 6]
```diff
@@ -0,0 +1,363 @@
+base_images:
+  aro-hcp-e2e-tests:
+    name: aro-hcp-e2e-tests
+    namespace: aro-hcp
+    tag: latest
+build_root:
+  project_image:
+    dockerfile_literal: |
```
Comment (deads2k) [inline ci-operator/config/Azure/ARO-HCP/Azure-ARO-HCP-main__capz-e2e.yaml:8]: ask test-platform if there is a better way. This is huge.

File: ci-operator/jobs/Azure/ARO-HCP/Azure-ARO-HCP-main-presubmits.yaml [hunk 7]
```diff
@@ -381,32 +389,39 @@ presubmits:
       - name: result-aggregator
         secret:
           secretName: result-aggregator
-    trigger: (?m)^/test( | .* )config-change-detection,?($|\s.*)
+    trigger: (?m)^/test( | .* )capz-e2e-dev,?($|\s.*)
   - agent: kubernetes
     always_run: true
```
Comment (deads2k) [inline ci-operator/jobs/Azure/ARO-HCP/Azure-ARO-HCP-main-presubmits.yaml]: start as optional and not by default

**Comment by deads2k** [comment]
some comments.  Be careful, be sure that you have cleanup in place for cancelled (jobs that are interrupted and never finish) BEFORE you merge.

/approve

**Comment by marek-veber** [comment]
/pj-rehearse periodic-ci-Azure-ARO-HCP-main-capz-e2e-production

File: ci-operator/jobs/Azure/ARO-HCP/Azure-ARO-HCP-main-presubmits.yaml (hunk 7 above)
Comment (marek-veber) [reply ci-operator/jobs/Azure/ARO-HCP/Azure-ARO-HCP-main-presubmits.yaml]: https://github.com/openshift/release/pull/80110/changes/bc6cff33cd90abe38a0ba4bf0c3922844f1a6f53
ok?

File: ci-operator/config/Azure/ARO-HCP/Azure-ARO-HCP-main__capz-e2e.yaml (hunk 6 above)
Comment (marek-veber) [reply ci-operator/config/Azure/ARO-HCP/Azure-ARO-HCP-main__capz-e2e.yaml:8]: After a suggestion from test-platform, I moved the env script generation from build_root into a dedicated write-env step — see [2491d40](https://github.com/openshift/release/pull/80110/changes/2491d40fd9dc586fda05e48d018164b643ae1800). The dockerfile_literal went from ~180 to ~60 lines. Is it better now?

Both tests pass with the refactored config:
* DEV — https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/80110/rehearse-80110-pull-ci-Azure-ARO-HCP-main-capz-e2e-dev/2067240653238046720
* PROD — https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/80110/rehearse-80110-periodic-ci-Azure-ARO-HCP-main-capz-e2e-production/2067000189360541696

For the DEV we have to solve the issue with unique ids (computed by ASO based on k8s names) - see   https://github.com/stolostron/cluster-api-installer/pull/701

**Comment by marek-veber** [comment]
> some comments. Be careful, be sure that you have cleanup in place for cancelled (jobs that are interrupted and never finish) BEFORE you merge.
> 
> /approve

Thank you for the approval! Regarding cleanup for cancelled/interrupted jobs:
* @RadekCap spent significant time tuning the teardown in the capi-tests framework.
* `allow_best_effort_post_steps: true` is set on both tests, so post steps (capz-arohcp-teardown → capz-arohcp-aks-deprovision) run even on failure.
* For DEV, we follow the same cleanup chain as the ARO-HCP base tests (aro-hcp-deprovision-environment).
* For PROD/STG/INT/DEV, capz-arohcp-aks-deprovision deletes the entire AKS resource group, which removes all child resources in one call.
* For the edge case where a job is hard-cancelled before post steps complete (we can add a periodic sweeper as a follow-up if needed):
  * The AKS resource groups have a predictable naming pattern (`${NAMESPACE}-${UNIQUE_HASH}-capz-rg`)
  * The ARO-HCP resources provisioned by CAPZ also have a predictable naming pattern (`capz-tests-XXXXXX-resgroup`, where `XXXXXX` is a random hex suffix)

File: ci-operator/step-registry/capz/test/summary/capz-test-summary-commands.sh [hunk 8]
```diff
@@ -4,7 +4,7 @@ set -o nounset
 set -o pipefail
 set -o xtrace
 
-source openshift-ci/capz-test-env.sh
+source "${SHARED_DIR}/capz-test-env.sh"
```
Comment (RadekCap) [inline ci-operator/step-registry/capz/test/summary/capz-test-summary-commands.sh]: The original `openshift-ci/capz-test-env.sh` (from the stolostron/capi-tests repo) had internal `{ set +o xtrace; } 2>/dev/null` guards around every credential read, so `set -o xtrace` in this script was safe — secrets were never expanded under tracing.

The new `${SHARED_DIR}/capz-test-env.sh` written by the `write-env` step contains literal `export AZURE_CLIENT_SECRET="<value>"` lines without those internal guards. With xtrace active, bash will print the expanded secret values to stderr, which ends up in CI build logs.

To preserve diagnostic tracing for `cp`, `make summary`, etc. while keeping secrets silent:

```suggestion
set -o xtrace

{ set +o xtrace; } 2>/dev/null
source "${SHARED_DIR}/capz-test-env.sh"
set -o xtrace
```

Comment (RadekCap) [reply ci-operator/step-registry/capz/test/summary/capz-test-summary-commands.sh]: A false positive. From logs.
`++ export AZURE_SUBSCRIPTION_ID=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX`

**Comment by marek-veber** [comment]
Pls.  [release-controller/OWNERS](https://github.com/openshift/release/blob/main/clusters/app.ci/release-controller/OWNERS):  @derekwaynecarr, @eparis, @jwforres, @dptp, @sosiouxme, @jupierce, @bradmwilliams

We have `/lgtm` from @RadekCap  & `/approve` from @deads2k 

We need from release-controller/OWNERS:
```
/approve
/label rehearsals-ack
```

Pls. help.

**Comment by bradmwilliams** [comment]
This file: `clusters/app.ci/release-controller/admin_01_releasepayload_crd.yaml` should not be included in this PR as it does not pertain to anything being included therein.  Frankly, I'd recommend rebasing against the latest/greatest code as these changes are already present at `main` and confirming that a `make release-controllers` command doesn't actually include this update at all.

**Comment by marek-veber** [comment]
/pj-rehearse periodic-ci-Azure-ARO-HCP-main-capz-e2e-production

**Comment by marek-veber** [comment]
/pj-rehearse periodic-ci-Azure-ARO-HCP-main-capz-e2e-production

**Comment by marek-veber** [comment]
> Rebase and ensure that `clusters/app.ci/release-controller/admin_01_releasepayload_crd.yaml` is not part of this PR

1. Rebase — done (rebased onto latest upstream/main)
2. Ensure admin_01_releasepayload_crd.yaml is not part of the PR — confirmed, it's not in any of our commits

**Comment by marek-veber** [comment]
@bradmwilliams - I did the rebase and the check - https://github.com/openshift/release/pull/80110#issuecomment-4745644744
